core/genesis.go

```diff
@@ -330,7 +330,10 @@ func SetupGenesisBlockWithOverride(db ethdb.Database, triedb *trie.Database, gen
 	}
 	// Just commit the new block if there is no stored genesis block.
 	stored := rawdb.ReadCanonicalHash(db, 0)
-	if (stored == common.Hash{}) {
+	// TODO: remove `true`. The problem is that for test consumption, the genesis block is stored
+	// in the db and is unaware of preimage recording. This means that we won't enter this if and
+	// save the genesis block state with preimages enabled, thus missing the genesis preimages.
+	if true || (stored == common.Hash{}) {
```

Comment:
Let me explain what's happening here, but first some context.

The way preimages work now is similar-ish to how `--cache.preimage` work, which setup the `Preimage` flag to create a new blockchain. This happens [here](https://github.com/gballet/go-ethereum/pull/466/files#diff-988a8d6651837931a44ef9d2b2fa45227543969e34fae4cd315220c3b2445b95R122) (i.e: only enabled for tests). For now, all fine.

But, just [some lines before that one](https://github.com/gballet/go-ethereum/pull/466/files#diff-988a8d6651837931a44ef9d2b2fa45227543969e34fae4cd315220c3b2445b95R112) we do `genesis.MustCommit(...)`. Under the hood this calls `genesis.Commit(...)`, and see the comment [here](https://github.com/gballet/go-ethereum/blob/kaustinen-with-shapella/core/genesis.go#L524-L526).

So the situation is that _for tests_, the genesis block is committed to the database without preimage recording, since now preimage recording is activated a bit later for the "blockchain setup".

When we do this blockchain setup (i.e: the current `SetupGenesisBlockWithOverride(...)` call we're in the current line), we check that the genesis block isn't already stored. If it is, we skip it (since "makes no sense" to do double work).

But... when we saved the genesis block in the `genesis.Commit(...)` it was unaware if we wanted to save preimages or not, so when that was generated and saved in the db there was no preimage recording.

What this means is that here we won't regenerate the gensis block again, so those preimages will always be missing. This `if true ||` ignores if the genesis block is in the db, and simply regenerates it again (but since now we're aware of the preimages flag, the preimages will be saved).

I did this change now just to show the situation. I can imagine multiple options:
- The `genesis.MustCommit(...)` receives the preimages flag and tries not saving the genesis block in the db, but that looks a bit hairy... since we need to add new parameters and doesn't feel the right place to do this.
- We can try to be able to detect in the current line if the `db` config has preimages activated, and change this if to `if db.config.Preimages || ...` which at least would recompute the genesis only if preimages were activated. But still, this feels wrongish.
- Remove the if and always recompute the genesis block. Sounds like redundant work.

The fundamental problem is that the testing code when running tests saves the genesis block into the db (maybe for a good reason) but it's unaware of preimages (nor has the logic to be able to deal saving them, as we have here).

I'll stop here since probably the situation is clear now, LMK your thoughts